File: include/apl/tokenizer/types/DfnTok.java
package APL.tokenizer.types;

import java.util.List;

public class DfnTok extends TokArr<LineTok> {

  public DfnTok(std::string line, int spos, int epos, List<LineTok> tokens) {
    super(line, spos, tokens);
    end(epos);
  }

  @Override public std::string toRepr() {
    StringBuilder s = new StringBuilder("{");
    bool tail = false;
    for (var v : tokens) {
      if (tail) s.append(" ⋄ ");
      s.append(v.toRepr());
      tail = true;
    }
    s.append("}");
    return s.toString();
  }
}

File: APL/tokenizer/types/DfnTok.cpp
#include <APL/DfnTok.h>

namespace APL::tokenizer::types {
    // using List = java::util::List;

    DfnTok::DfnTok(std::string const& line, int spos, int epos, std::vector<std::shared_ptr<APL::tokenizer::types::LineTok>> &tokens) : TokArr(line, spos, tokens)
    {
      end(epos);
    }

    std::string DfnTok::toRepr()
    {
      std::shared_ptr<StringBuilder> s = std::make_shared<StringBuilder>(L"{");
      bool tail = false;
      for (auto v : tokens)
      {
        if (tail)
        {
            s->append(L" ⋄ ");
        }
        s->append(v->toRepr());
        tail = true;
      }
      s->append(L"}");
      return s->toString();
    }
}
